### DISCUSSION TITLE
fix: How to gracefully join worker thread when rocket shuts down?

### DISCUSSION THREAD

**Discussion by gabyx** [rwf2/Rocket#2711] [Questions]
I need a worker thread along `rocket` which should shutdown gracefully when rocket shuts down.
But the thing does not compile because `Fairing::on_shutdown` is taking `&self` instead of `&mut self`.

I come [up with this](https://www.rustexplorer.com/b#LyoKW2RlcGVuZGVuY2llc10KW2RlcGVuZGVuY2llcy5yb2NrZXRdCnZlcnNpb24gPSAiPTAuNS4wIgpmZWF0dXJlcyA9IFsnanNvbiddCiovCgp1c2Ugcm9ja2V0OjpmYWlyaW5nOjp7RmFpcmluZywgSW5mbywgS2luZH07CnVzZSBzdGQ6OnsKICAgIHN5bmM6OnsKICAgICAgICBtcHNjOjp7c2VsZiwgU2VuZGVyLCBUcnlSZWN2RXJyb3J9LAogICAgfSwKICAgIHRocmVhZDo6e3NlbGYsIEpvaW5IYW5kbGV9LAogICAgdGltZTo6RHVyYXRpb24sCn07CgplbnVtIE5vdGlmaWNhdGlvbiB7CiAgICBDYW5jZWwsCn0KCnN0cnVjdCBTdGF0dXNDb25zdW1lclNodXRkb3duIHsKICAgIGhhbmRsZTogSm9pbkhhbmRsZTwoKT4sCiAgICB0eDogU2VuZGVyPE5vdGlmaWNhdGlvbj4sCn0KCiNbcm9ja2V0Ojphc3luY190cmFpdF0KaW1wbCBGYWlyaW5nIGZvciBTdGF0dXNDb25zdW1lclNodXRkb3duIHsKICAgIGZuIGluZm8oJnNlbGYpIC0+IEluZm8gewogICAgICAgIEluZm8gewogICAgICAgICAgICBuYW1lOiAiU3RhdHVzIENvbnN1bWVyIFNodXRkb3duIEZhaXJpbmciLAogICAgICAgICAgICBraW5kOiBLaW5kOjpTaHV0ZG93biwKICAgICAgICB9CiAgICB9CgogICAgLy8vIENhbm5vdCBjaGFuZ2UgdGhpcyB0cmFpdCB0byBgc2VsZmAuLi4uLi4gCiAgICBhc3luYyBmbiBvbl9zaHV0ZG93bigmc2VsZiwgX3JvY2tldDogJnJvY2tldDo6Um9ja2V0PHJvY2tldDo6T3JiaXQ+KSB7CiAgICAgICAgaWYgbGV0IEVycihlKSA9IHNlbGYudHguc2VuZChOb3RpZmljYXRpb246OkNhbmNlbCkgewogICAgICAgICAgICBwcmludGxuISgKICAgICAgICAgICAgICAgICJDb3VsZCBub3Qgc2VuZCBjYW5jZWwgbm90aWZpY2F0aW9uOiBlcnJvcjpcbnt9IiwgZQogICAgICAgICAgICApOwogICAgICAgIH0KCiAgICAgICAgd2hpbGUgIXNlbGYuaGFuZGxlLmlzX2ZpbmlzaGVkKCkgewogICAgICAgICAgICBzZWxmLmhhbmRsZS5qb2luKCk7CiAgICAgICAgfQogICAgfQp9CgpwdWIgZm4gc3Bhd25fc3RhdHVzX2NvbnN1bWVyKCkgLT4gaW1wbCBGYWlyaW5nIHsKICAgIGxldCAodHgsIHJ4KSA9IG1wc2M6OmNoYW5uZWw6OjxOb3RpZmljYXRpb24+KCk7CgogICAgbGV0IGhhbmRsZSA9IHRocmVhZDo6c3Bhd24obW92ZSB8fCBsb29wIHsKICAgICAgICBwcmludGxuISgiV29ya2luZy4uLiIpOwogICAgICAgIHRocmVhZDo6c2xlZXAoRHVyYXRpb246OmZyb21fbWlsbGlzKDUwMCkpOwoKICAgICAgICBtYXRjaCByeC50cnlfcmVjdigpIHsKICAgICAgICAgICAgT2soTm90aWZpY2F0aW9uOjpDYW5jZWwpID0+IHsKICAgICAgICAgICAgICAgIHByaW50bG4hKCJSZWNlaXZlZCBjYW5jZWxhdGlvbiAtPiBUZXJtaW5hdGluZy4iKTsKICAgICAgICAgICAgICAgIGJyZWFrOwogICAgICAgICAgICB9CiAgICAgICAgICAgIEVycihUcnlSZWN2RXJyb3I6OkRpc2Nvbm5lY3RlZCkgPT4gewogICAgICAgICAgICAgICAgcHJpbnRsbiEoICJQYXJlbnQgdGhyZWFkIGRpc2Nvbm5lY3RlZCAtPiBUZXJtaW5hdGluZy4iKTsKICAgICAgICAgICAgICAgIGJyZWFrOwogICAgICAgICAgICB9CiAgICAgICAgICAgIEVycihUcnlSZWN2RXJyb3I6OkVtcHR5KSA9PiB7fQogICAgICAgIH0KICAgIH0pOwoKICAgIHJldHVybiBTdGF0dXNDb25zdW1lclNodXRkb3duIHsgaGFuZGxlLCB0eCB9Owp9CgpmbiBtYWluKCkgewogICAgbGV0IHNodXRkb3duX2ZhaXJpbmcgPSBzcGF3bl9zdGF0dXNfY29uc3VtZXIoKTsKICAgIHJvY2tldDo6YnVpbGQoKS5hdHRhY2goc2h1dGRvd25fZmFpcmluZykubGF1bmNoKCk7Cn0=)

```rust
/*
[dependencies]
[dependencies.rocket]
version = "=0.5.0"
features = ['json']
*/

use rocket::fairing::{Fairing, Info, Kind};
use std::{
    sync::{
        mpsc::{self, Sender, TryRecvError},
    },
    thread::{self, JoinHandle},
    time::Duration,
};

enum Notification {
    Cancel,
}

struct StatusConsumerShutdown {
    handle: JoinHandle<()>,
    tx: Sender<Notification>,
}

#[rocket::async_trait]
impl Fairing for StatusConsumerShutdown {
    fn info(&self) -> Info {
        Info {
            name: "Status Consumer Shutdown Fairing",
            kind: Kind::Shutdown,
        }
    }

    /// Cannot change this trait to `self`...... 
    async fn on_shutdown(&self, _rocket: &rocket::Rocket<rocket::Orbit>) {
        if let Err(e) = self.tx.send(Notification::Cancel) {
            println!(
                "Could not send cancel notification: error:\n{}", e
            );
        }

        /// I know that the following loop is blocking the executor, 
        /// but that is not relevant for the question.
        while !self.handle.is_finished() {
            self.handle.join(); ///////////// DOES NOT COMPILE DUE TO `&self`.
        }
    }
}

pub fn spawn_status_consumer() -> impl Fairing {
    let (tx, rx) = mpsc::channel::<Notification>();

    let handle = thread::spawn(move || loop {
        println!("Working...");
        thread::sleep(Duration::from_millis(500));

        match rx.try_recv() {
            Ok(Notification::Cancel) => {
                println!("Received cancelation -> Terminating.");
                break;
            }
            Err(TryRecvError::Disconnected) => {
                println!( "Parent thread disconnected -> Terminating.");
                break;
            }
            Err(TryRecvError::Empty) => {}
        }
    });

    return StatusConsumerShutdown { handle, tx };
}

fn main() {
    let shutdown_fairing = spawn_status_consumer();
    rocket::build().attach(shutdown_fairing).launch();
}
```

This is a cross post, and the question is basically here:

https://users.rust-lang.org/t/how-to-have-joinhandle-inside-rocket-fairing-for-shutdown/105748/2

Thanks a lot for the help.

**Comment by SergioBenitez** [comment]
Is there a reason you're starting a blocking thread as opposed to a tokio task? If you did the latter, the problem would be simplified. Something like the following, in pseudocode:

```rust
impl Fairing for SomeFairing {
    async fn on_shutdown(&self) -> _ {
        self.rx.await;
    }
}

#[launch]
fn rocket() -> _ {
    let (tx, rx) = tokio::sync::oneshot::channel();
    let rocket = rocket::build()
        .mount("/", routes![...])
        .attach(SomeFairing { rx });

    let shutdown = rocket.shutdown();
    tokio::task::spawn(async move {
        loop {
            select! {
                _ = shutdown => break,
                msg = channel => do(msg),
            }
        }

        cleanup().await;
        tx.send(());
    });

    rocket
}
```

If you must use a regular thread, you can do _almost_ the same thing, but keep your cancellation channel around and use it to send the shutdown message instead of using the `Shutdown` future.

**Reply by gabyx** [reply]
What do you mean with blocking thread? The thread is not blocking, it gets spawn (yes blocking the executor, because not async...) and then it should be along side rocket.
I just didn't want to use "green" threads (rocket tasks (async)). Because I wanted to move out of async world for this task.
I will consider it. Async is like a curse, once you have it it propells everywhere. =).

What does `let shutdown = rocket.shutdown();` do?

**Reply by SergioBenitez** [reply]
https://api.rocket.rs/v0.5/rocket/struct.Rocket.html#method.shutdown-1

**Reply by gabyx** [reply]
The solution troubles me since `shutdown` inside the `loop` is not allowed:

```rust
 |
65 |                 _ = shutdown => break,
   |                     ^^^^^^^^ value moved here, in previous iteration of loop
   |
   = note: move occurs because `shutdown` has type `rocket::Shutdown`, which does not implement the `Copy` trait
```

what I am doing wrong?

**Reply by gabyx** [reply]
```rust
pub struct WaitForShutdown {
    pub shutdown: Shutdown,
    pub tx_consumer_done: Sender<()>,
}


/// Spawn a status queue consumer thread.
pub async fn spawn_status_consumer(
    l: Arc<Logger>,
    queue: StatusQueue,
    database_url: &str,
    wait_for_shutdown: WaitForShutdown,
) {
    let db_conn = database::connect(&l, &database_url);

    let log = l.clone();
    let shutdown = wait_for_shutdown.shutdown;

    tokio::task::spawn(async move {
        loop {
            select! {
                _ = shutdown => break,
                _ = async {
                    info!(log, "Working ...");
                    tokio::time::sleep(Duration::from_millis(100)).await;
                } => ()
            }
        }

        // Notify that we finished.
        wait_for_shutdown.tx_consumer_done.send(());
    });
}
```

**Reply by gabyx** [reply]
Ok a `shutdown.clone()` has helped. 
[The other problem is: ](https://www.rustexplorer.com/b#LyoKW2RlcGVuZGVuY2llc10KW2RlcGVuZGVuY2llcy5yb2NrZXRdCnZlcnNpb24gPSAiPTAuNS4wIgpmZWF0dXJlcyA9IFsnanNvbiddCiovCgp1c2Ugcm9ja2V0Ojp7CiAgICBmYWlyaW5nOjp7RmFpcmluZywgSW5mbywgS2luZH0sCiAgICB0b2tpbzo6ewogICAgICAgIHNlbGYsIHNlbGVjdCwKICAgICAgICBzeW5jOjpvbmVzaG90Ojp7UmVjZWl2ZXIsIFNlbmRlcn0sCiAgICB9LAogICAgU2h1dGRvd24sCn07CnVzZSBzdGQ6Ont0aW1lOjpEdXJhdGlvbn07CgplbnVtIE5vdGlmaWNhdGlvbiB7CiAgICBDYW5jZWwsCn0KCi8vLyBSb2NrZXQgZmFpcmluZyB0byB3YWl0IGZvciBjb25zdW1lciB0byBiZSBkb25lLgpwdWIgc3RydWN0IFdhaXRGb3JDb25zdW1lckRvbmUgewogICAgcHViIHJ4X2NvbnN1bWVyX2RvbmU6IFJlY2VpdmVyPCgpPiwKfQoKLy8vIEhlbHBlciBkYXRhIGZvciB0aGUgc3RhdHVzIGNvbnN1bWVyIHRvIHJlYWN0IG9uIGEgc2h1dGRvd24gYW5kIHRvIGFja25vd2xlZGdlIHRoaXMKLy8vIGJ5IGB0eF9zaHV0ZG93bmAuCnB1YiBzdHJ1Y3QgV2FpdEZvclNodXRkb3duIHsKICAgIHB1YiBzaHV0ZG93bjogU2h1dGRvd24sCiAgICBwdWIgdHhfY29uc3VtZXJfZG9uZTogU2VuZGVyPCgpPiwKfQoKI1tyb2NrZXQ6OmFzeW5jX3RyYWl0XQppbXBsIEZhaXJpbmcgZm9yIFdhaXRGb3JDb25zdW1lckRvbmUgewogICAgZm4gaW5mbygmc2VsZikgLT4gSW5mbyB7CiAgICAgICAgSW5mbyB7CiAgICAgICAgICAgIG5hbWU6ICJTdGF0dXMgQ29uc3VtZXIgU2h1dGRvd24gRmFpcmluZyIsCiAgICAgICAgICAgIGtpbmQ6IEtpbmQ6OlNodXRkb3duLAogICAgICAgIH0KICAgIH0KCiAgICBhc3luYyBmbiBvbl9zaHV0ZG93bigmc2VsZiwgX3JvY2tldDogJnJvY2tldDo6Um9ja2V0PHJvY2tldDo6T3JiaXQ+KSB7CiAgICAgICAgc2VsZi5yeF9jb25zdW1lcl9kb25lLmF3YWl0OwogICAgfQp9CgovLy8gU3Bhd24gYSBzdGF0dXMgcXVldWUgY29uc3VtZXIgdGhyZWFkLgpwdWIgYXN5bmMgZm4gc3Bhd25fc3RhdHVzX2NvbnN1bWVyKAogICAgd2FpdF9mb3Jfc2h1dGRvd246IFdhaXRGb3JTaHV0ZG93biwKKSB7CiAgICB0b2tpbzo6dGFzazo6c3Bhd24oYXN5bmMgbW92ZSB7CiAgICAgICAgbG9vcCB7CiAgICAgICAgICAgIHNlbGVjdCEgewogICAgICAgICAgICAgICAgXyA9IHdhaXRfZm9yX3NodXRkb3duLnNodXRkb3duLmNsb25lKCkgPT4gYnJlYWssCiAgICAgICAgICAgICAgICBfID0gYXN5bmMgewogICAgICAgICAgICAgICAgICAgIHRva2lvOjp0aW1lOjpzbGVlcChEdXJhdGlvbjo6ZnJvbV9taWxsaXMoMTAwKSkuYXdhaXQ7CiAgICAgICAgICAgICAgICB9ID0+ICgpCiAgICAgICAgICAgIH0KICAgICAgICB9CgogICAgICAgIC8vIE5vdGlmeSB0aGF0IHdlIGZpbmlzaGVkLgogICAgICAgIHdhaXRfZm9yX3NodXRkb3duLnR4X2NvbnN1bWVyX2RvbmUuc2VuZCgoKSk7CiAgICB9KTsKfQoKZm4gbWFpbigpIHsKfQ==)

![image](https://github.com/rwf2/Rocket/assets/647437/a945de03-54cf-4ed3-825d-8d6afd56804f)

**Reply by SergioBenitez** [reply]
It feels like we're trying to solve some problem some specific way because that's the way you tried to solve the problem initially, but there may be an entirely different approach at hand. What is the actual thing you're trying to do? Based on the names in your source code, it sounds like there's some "status consumer"? What does it do, and why does it need to be notified of shutdown?

**Reply by gabyx** [reply]
I had a `status consumer` (a green thread next to rocket which consumes from a RabbitMQ queue) and when Rocket gets killed in  kubernetes by s Signal, I just want that this status consumer gracefully joins  before Rocket asyn main is resolved and the main function exits, which tears down the `status consumer` (it might be in the middle of writing to a DB ...). So all I wanted it gracefull termination. I solved the above by doing [this](https://github.com/gabyx/RsMarkdown2PDF-Service/blob/main/components/api/src/service/consumer.rs#L78). Which installs a consumer in the RabbitMQ queue then spawns a async listener for `shutdown` and when it happens it will unsubscribe the consumer (not sure if this await also waits till any consumer callback is finished) and then notifies the over the transmitt channel that the consumer green-thread has finished... Not sure if everything is right, but it works...
